File: Leetcode/Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char>  matchingPair;
        matchingPair[')'] = '(';matchingPair['}'] = '{';
        matchingPair[']'] = '[';

        stack<char> opennings;

        for(int i = 0;i < s.size();i++){
            if(isOpenningBracket(s[i])){
                opennings.push(s[i]);
            }
            else{
                if(    opennings.empty()
                    || opennings.top() != matchingPair[s[i]])

                    return false;

                opennings.pop();
            }
        }

        return opennings.empty();
    }

    inline bool isOpenningBracket(char c){
        return c == '(' || c == '{' || c == '[';
    }
};

File: Leetcode/Valid Parentheses.cs
public class Solution {
    public bool IsValid(string str) {
        Stack<char> s = new Stack<char>();
        
        Dictionary<char,char> pair = new Dictionary<char,char>(){
            {')','('},
            {'}','{'},
            {']','['},
        } ;
        
        foreach(char c in str)
        {
            if(c == '(' || c == '{' || c == '[')
                s.Push(c);
            else if(s.Count == 0 || s.Pop() != pair[c])
                    return false;
        }
        return s.Count == 0;
    }
}